### ISSUE TITLE
Some suggestions to enhance locating logics -'offset()' and 'resize()'

### ISSUE THREAD

**Issue by forhonourlx** [jstockwin/py-pdf-parser#111]
Hi @jstockwin ，

I am coming with some suggestions from my recent practice, would you please take a look? Thanks in advance.
1.  Adding 'offset()' and 'resize()' shortcut method of ElementList also Element:
For `ElementList.between()` method, with `inclusive` param, users are only able to select elements like `(e_1, ...e_n)` or `[e_1, ..., e_n]`. In first case, I want to select elements like `(e_1, ...e_n]` or `[e_1, ..., e_n)`(i.e.). In further case, I want to select elements like `(e_1-3, ...e_n+4)`, which requires elements outside the 'between' elements, maybe also requires `before`, `after` slices.
So I suggest giving users the freedom to easily choose elements:
'offset()': Giving e_0, I am able to choose e_-3 or e_3, so that I can easily locating elements between 2 offseted elements.
'resize()': Giving an element e_0, I am able to choose [e_0,...e_4].

The following is some demo of my actual usage without offset and resize, which troubles me.
```
    def locate_contents_between_locs(self, loc_pair_text,
            loc_pair_equal_or_contain=['equal','equal'],
            loc_pair_idx=[0,0],
            loc_pair_offset=[0,0],
            elements_scope=None,
            output_type=list,
            if_print_contents=False,
            ):
        if elements_scope == None:
            elements = self.doc.elements
        else:
            elements = elements_scope

        loc_pair = [None, None]
        for _ in range(2):
            if loc_pair_equal_or_contain[_] =='equal':
                loc_pair[_] = elements.filter_by_text_equal(loc_pair_text[_])[loc_pair_idx[_]]
            elif loc_pair_equal_or_contain[_] =='contain':
                loc_pair[_] = elements.filter_by_text_contains(loc_pair_text[_])[loc_pair_idx[_]]

            if loc_pair_offset[_] > 0:#inclusive [0, 1, 2...
                loc_pair[_] = elements.after(loc_pair[_], inclusive=True)[loc_pair_offset[_]]
            elif loc_pair_offset[_] < 0:#exclusive ...-2, -1]
                loc_pair[_] = elements.before(loc_pair[_], inclusive=False)[loc_pair_offset[_]]

        contents = elements.between(loc_pair[0], loc_pair[1])
        if if_print_contents:
            print(f'{contents}:{[c.text() for c in contents]}')
        
        return contents
```

**Comment by forhonourlx** [comment]
Also it is an idea of locating elements by relative index and element count.

**Comment by jstockwin** [comment]
Hi @forhonourlx, thanks for taking the time to write such a detailed issue - that's really helpful.

I do have some ideas for how this would be implemented, but would you mind first telling me a bit more about your use case?

My main points are:

- py-pdf-parser is aimed at making it easy to extract certain elements and certain information. Why the need to move from `e_0` to `e_4`? Why can't you just locate `e_4` in the first place?

- The `offset` function you're requesting is fairly simple to code work around using the existing API (i.e. `ElementList.offset(element, count)` is the same as `ElementList.after(element, inclusive=True)[count]`). Whilst I appreciate you're having to play around with `after` or `before`, and also `inclusive`, I'd want a solid use case before implementing a specific function to do this since.

- Your `resize` function is trivial given `offset`, so I think in that case you'd just call `ElementList.between(element, ElementList.ofset(element, count))`.

To summarise, I do have some ideas for adding an `offset` type function, but given how easily this functionality is achieved with the current API, I'd like a solid use case before adding it, since adding it creates more code to maintain.

**Comment by forhonourlx** [comment]
Hi @jstockwin,

For privacy reason, I cannot publish an actual pdf example. So I am trying to explain in words:
> Why the need to move from e_0 to e_4? Why can't you just locate e_4 in the first place?

A: In some frequent cases, I can only locate `e0`(like a header or title) at first place and  e_4 is not a certain criteria(but certain by element count). Sometimes, there maybe some annoying multi-index header or anything I don't want, so that I can skip certain count of elements or rows by element count(e.g. 2 elements in row0, and 3 in row1 to skip, now at e_5) to start extracting tables.

> The offset function you're requesting is fairly simple to code work around using the existing API (i.e. ElementList.offset(element, count) is the same as ElementList.after(element, inclusive=True)[count])
Your resize function is trivial given offset...

A: Yes, it could work with existing API, but too many steps to implement.
For example, to get an ElementList from e_-4 to e_n+3: 
① `between` 2 searched criteria(e_0, e_n), ② `after` e_n 3 count, get e_n+3; ③`before` e_0 4 count, get e_-4; ④ `between` again with e_-4 and e_n+3 to get final contents.

My ideal methods are quite simple: 
I. ① `between` 2 searched criteria(e_0, e_n) , get element_list; ②element_list.resize(3).resize(-4) to get contents
II. ① `offset` 2 searched criteria, get `e_-4 = e_0.offset(-4)`, `e_n+3 = e_n.offset(3)`; ②`between` e_-4 and e_n+3 to get contents

In conlusion, it would be nice to give users easy, friendly, free ways to slice ElementList. Text criteria  may be found at title, header, middle, tail, next title or anywhere, `between` with symmetric `inclusive` or exclusive is not flexible and convient enough to deal with some frequent cases. In my practice, I found it quite useful and accurate to locate contents with text criteria and relative count.

**Comment by jstockwin** [comment]
@forhonourlx Alright, thanks for the explanation.

In general, I would encourage you to look for other ways of locating these elements. For example, make use of some of the geometric functions `to_the_left_of` or `vertically_in_line_with`. You can also do set operations on element lists, so for example `elements.below(e_1).to_the_left_of(e_2)`.

That said, I can see some scope to help with your ideal method II. What do you think of the following?

Rather than `offset` on an element, I'd add `next` to the element list. This would be called with the start element, and a count (which defaults to 1). So you can do `ElementList.next(element, count)`. If count is positive it moves you forward in the list, and if it's negative it moves you backward. I would also add `previous` which does the reverse.

In your case, you would then be able to do:
```
e_neg_4 = element_list.previous(e_0, 4)
e_n_plus_3 = element_list.next(e_n, 3)
contents = element_list.between(e_neg_4, e_n_plus_3)
```

I'm not a fan of resizing element lists - you should just instantiate them with the right elements in the first place. However, hopefully, you don't need this given the above.

What do you think?

**Comment by jstockwin** [comment]
Also, you mention that "`between` with symmetric `inclusive` or exclusive is not flexible and convient enough[...]".

Note that if you want different inclusivity at each end, you can simply chain `before` and `after`, i.e. `elements.after(e_0, inclusive=True).before(e_n, inclusive=False)` gives you `between(e_0, e_n)` with different inclusivity at each end.

**Comment by forhonourlx** [comment]
Hi @jstockwin ,

`elements.after(e_0, inclusive=True).before(e_n, inclusive=False)` is more flexible than `between`.  Do I miss the chain in the documentation examples? I will take more practice in `chain` .

`ElementList.next(element, count)` is identical to `offset` idea, I really appreciate it if adding this.

I would raise more suggestions during my practice, hoping not bothering you. Thanks!

**Comment by jstockwin** [comment]
Chaining is mentioned in the [filtering overview](https://py-pdf-parser.readthedocs.io/en/latest/overview.html#filtering) or the docs, although I'm not sure I use it in any examples.

I agree it's an identical idea. No problem, I'll get round to adding that soon.

> I would raise more suggestions during my practice

No problem at all! It's interesting to hear how you're using it.

**Comment by jstockwin** [comment]
As an additional note - once I started working on this I realised `next` is not a good name, since this is a python builtin.

I've decided to name the functions very explicitly. I think this is also good since most of the functions on an `ElementList` are for filtering - it's in `filtering.py`, and this is how it's referenced in the docs. These new functions do not filter, so I think it's good that their name makes it clear that this is not the case.

I'm going with `move_forwards_from` and `move_backwards_from`.

**Comment by jstockwin** [comment]
@forhonourlx This is done, and released in version `0.5.0` which should be on `PyPI` shortly.
